{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3 as sq1\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  assession            seq\n",
      "0  AY911262  ATGGACTCTTACT\n",
      "1  FJ614813  CTGGACGGTTTCC\n",
      "2  FJ948820  GCTTTCGGGATAC\n",
      "3  JF920046  ATGGACTCTTACT\n",
      "4  JF920047  GCTTTCGGGATAC\n",
      "  assession            seq     journal\n",
      "0  AY911262  ATGGACTCTTACT    J. Virol\n",
      "1  FJ614813  CTGGACGGTTTCC    J. Virol\n",
      "2  JF920047  GCTTTCGGGATAC  PLoS ONE 6\n",
      "  assession            seq country\n",
      "0  AY911262  ATGGACTCTTACT     USA\n",
      "1  FJ614813  CTGGACGGTTTCC     USA\n",
      "2  JF920046  ATGGACTCTTACT     USA\n",
      "3  JF920047  GCTTTCGGGATAC     USA\n",
      "  assession            seq genotype\n",
      "0  AY911262  ATGGACTCTTACT      ON1\n",
      "1  FJ614813  CTGGACGGTTTCC      ON1\n",
      "2  FJ948820  GCTTTCGGGATAC      GA2\n",
      "3  JF920046  ATGGACTCTTACT       BA\n",
      "4  JF920047  GCTTTCGGGATAC      GB3\n",
      "  assession clade genotype  epitope\n",
      "0  AY911262    IV      GA2   CTGGAC\n",
      "1  FJ614813    II      GA2   CTGGAC\n",
      "2  FJ948820     I      GA2   CTGGAC\n",
      "3  JF920046     I       BA   AGATAC\n",
      "4  JF920047   III      GB3  GCTTTCG\n",
      "this excuted\n"
     ]
    }
   ],
   "source": [
    "conn =sq1.connect(\"RSV.db\")\n",
    "\n",
    "try:\n",
    "    conn.execute(\"\"\"\n",
    "    DROP TABLE IF EXISTS major_clade\n",
    "    \"\"\")\n",
    "    \n",
    "    conn.execute(\"\"\"\n",
    "    DROP TABLE IF EXISTS subclade\n",
    "    \"\"\")\n",
    "    \n",
    "    conn.execute(\"\"\"\n",
    "    DROP TABLE IF EXISTS sequence_epitope\n",
    "    \"\"\")\n",
    "    \n",
    "    conn.execute(\"\"\"\n",
    "    DROP TABLE IF EXISTS genebank\n",
    "    \"\"\")\n",
    "    \n",
    "\n",
    "    \n",
    "    conn.execute(\"\"\"\n",
    "    CREATE TABLE IF NOT EXISTS major_clade (subtype TEXT, genotype TEXT, PRIMARY KEY ('genotype'));\n",
    "    \"\"\")\n",
    "    conn.execute(\"\"\"\n",
    "    CREATE TABLE IF NOT EXISTS subclade (genotype TEXT, nucleotide TEXT, subclade INT\n",
    "                    FOREIGN KEY('genotype') \n",
    "                    REFERENCES major_clade ('genotype')\n",
    "                    ON DELETE CASCADE\n",
    "                    ON UPDATE CASCADE);\n",
    "    \"\"\")\n",
    "    conn.execute(\"\"\"\n",
    "    CREATE TABLE IF NOT EXISTS sequence_epitope (id TEXT, nucleotide TEXT, amino_acid TEXT, binding_score TEXT,\n",
    "                    PRIMARY KEY('id')                  \n",
    "                    ON DELETE CASCADE\n",
    "                    ON UPDATE CASCADE);  \n",
    "    \"\"\")\n",
    "    \n",
    "    conn.execute(\"\"\"\n",
    "    CREATE TABLE IF NOT EXISTS genbank (id TEXT, country TEXT, year INT,\n",
    "                    FOREIGN KEY('id') \n",
    "                    REFERENCES sequence_epitope ('id')\n",
    "                    ON DELETE CASCADE\n",
    "                    ON UPDATE CASCADE);  \n",
    "    \"\"\")\n",
    "    \n",
    "    \n",
    "    \n",
    "    conn.executemany(\"\"\"\n",
    "    INSERT INTO major_clade VALUES (?,?,?)\"\"\",[('A','AY911262','ON1'),\n",
    "                                               ('A'.'FJ614813','ON1'),\n",
    "                                               ('A','FJ948820','GA2'),\n",
    "                                               ('B','JF920046','BA'),\n",
    "                                               ('B','JF920047','GB3')])\n",
    "\n",
    "    \n",
    "    conn.executemany(\"\"\"\n",
    "    INSERT INTO subclade VALUES (?, ?)\"\"\",[('A','ATGGACTCTTACT'),\n",
    "                                           ('FJ614813','CTGGACGGTTTCC'),\n",
    "                                           ('FJ948820','GCTTTCGGGATAC'),\n",
    "                                           ('JF920046','ATGGACTCTTACT'),\n",
    "                                           ('JF920047','GCTTTCGGGATAC')])\n",
    "    conn.executemany(\"\"\"\n",
    "    INSERT INTO metadata VALUES (?,?,?)\"\"\",[('AY911262','USA',2007),\n",
    "                                           ('FJ614813','USA',2009),\n",
    "                                           ('FJ948820','Netherland',2010),\n",
    "                                           ('JF920046','USA',2012),\n",
    "                                           ('JF920047','USA',2009)])\n",
    "    conn.executemany(\"\"\"\n",
    "    INSERT INTO ref VALUES (?,?,?)\"\"\",[('AY911262','Respiratory syncytial virus nonstructural proteins NS1 and NS2','J. Virol'),\n",
    "                                      ('FJ614813','A chimeric A2 strain of respiratory syncytial virus','J. Virol'),\n",
    "                                      ('JF920047','Whole Genome Sequencing and Evolutionary','PLoS ONE 6')])\n",
    "    conn.executemany(\"\"\"\n",
    "    INSERT INTO phylogenetic VALUES (?,?,?)\"\"\",[('AY911262','ON1','IV'),\n",
    "                                               ('FJ614813','ON1','II'),\n",
    "                                               ('FJ948820','GA2','I'),\n",
    "                                               ('JF920046','BA','I'),\n",
    "                                               ('JF920047','GB3','III')])\n",
    "    \n",
    "    conn.executemany(\"\"\"\n",
    "    INSERT INTO epitope VALUES (?, ?)\"\"\",[('ON1','ATGGAC'),\n",
    "                                          ('GA2','CTGGAC'),\n",
    "                                          ('BA','AGATAC'),\n",
    "                                          ('GB3','GCTTTCG')])\n",
    "\n",
    "    cursor1 = conn.execute(\"select * from sequence\")\n",
    "    df1 = pd.DataFrame(cursor1.fetchall(),columns = [ 'assession','seq'] )              \n",
    "    print(df1)\n",
    "\n",
    "    \n",
    "    cursor2 = conn.execute('''select sequence.assession, sequence.seq, ref.journal from\n",
    "                           sequence join ref\n",
    "                           on sequence.assession = ref.id''')\n",
    "                           \n",
    "    df2 = pd.DataFrame(cursor2.fetchall(),columns = [ 'assession','seq','journal'] )              \n",
    "    print(df2)\n",
    "    \n",
    "    cursor3 = conn.execute('''select sequence.assession, sequence.seq, metadata.country from\n",
    "                           sequence join metadata\n",
    "                           on sequence.assession = metadata.assession\n",
    "                           where country = 'USA' ''')\n",
    "    df3 = pd.DataFrame(cursor3.fetchall(),columns = [ 'assession','seq','country'] )              \n",
    "    print(df3)\n",
    "    \n",
    "    cursor4 = conn.execute('''select sequence.assession, sequence.seq, phylogenetic.genotype from \n",
    "                           sequence join phylogenetic\n",
    "                           on sequence.assession = phylogenetic.assession\n",
    "                           order by sequence.assession asc''')\n",
    "    df4 = pd.DataFrame(cursor4.fetchall(),columns = [ 'assession','seq','genotype'] )              \n",
    "    print(df4)\n",
    "    \n",
    "    cursor5 = conn.execute('''update phylogenetic set genotype = 'GA2' where genotype = 'ON1';''')\n",
    "    cursor5 = conn.execute('''select phylogenetic.assession, phylogenetic.clade, phylogenetic.genotype,epitope.epitope from\n",
    "                           epitope join phylogenetic\n",
    "                           on phylogenetic.genotype = epitope.genotype''')\n",
    "    df5 = pd.DataFrame(cursor5.fetchall(),columns = [ 'assession','clade','genotype','epitope'] )              \n",
    "    print(df5)                      \n",
    "                           \n",
    "                           \n",
    "    conn.commit();\n",
    "                           \n",
    "                           \n",
    "except sqlite3.Error as e:\n",
    "    print(\"There was error:\"+str(e))   \n",
    "    \n",
    "finally: \n",
    "    print(\"this excuted\")\n",
    "    conn.close()                  \n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
