{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3 as sq1\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  assession            seq\n",
      "0  AY911262  ATGGACTCTTACT\n",
      "1  FJ614813  CTGGACGGTTTCC\n",
      "2  FJ948820  GCTTTCGGGATAC\n",
      "3  JF920046  ATGGACTCTTACT\n",
      "4  JF920047  GCTTTCGGGATAC\n",
      "  assession            seq  epitope\n",
      "0  AY911262  ATGGACTCTTACT   ATGGAC\n",
      "1  FJ614813  CTGGACGGTTTCC   CTGGAC\n",
      "2  FJ948820  GCTTTCGGGATAC   AGATAC\n",
      "3  JF920046  ATGGACTCTTACT  ATGGACT\n",
      "4  JF920047  GCTTTCGGGATAC  GCTTTCG\n",
      "  assession            seq country\n",
      "0  AY911262  ATGGACTCTTACT     USA\n",
      "1  FJ614813  CTGGACGGTTTCC     USA\n",
      "2  JF920046  ATGGACTCTTACT     USA\n",
      "3  JF920047  GCTTTCGGGATAC     USA\n",
      "  assession            seq genotype\n",
      "0  FJ614813  CTGGACGGTTTCC      ON1\n",
      "1  FJ948820  GCTTTCGGGATAC      GA2\n",
      "2  JF920046  ATGGACTCTTACT       BA\n",
      "3  JF920047  GCTTTCGGGATAC      GB3\n",
      "this excuted\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'sqlite3' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOperationalError\u001b[0m                          Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-661fdc4f4edb>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    114\u001b[0m                            \u001b[0msequence\u001b[0m \u001b[0mjoin\u001b[0m \u001b[0mphylogenetic\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 115\u001b[0;31m                            on sequence.assession = phylogenetic.assession''')\n\u001b[0m\u001b[1;32m    116\u001b[0m     \u001b[0mdf5\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcursor5\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfetchall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m \u001b[0;34m'assession'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'seq'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'country'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mOperationalError\u001b[0m: no such column: metadata.assession",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-661fdc4f4edb>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    121\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    122\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 123\u001b[0;31m \u001b[0;32mexcept\u001b[0m \u001b[0msqlite3\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    124\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"There was error:\"\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    125\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'sqlite3' is not defined"
     ]
    }
   ],
   "source": [
    "conn =sq1.connect(\"RSV.db\")\n",
    "\n",
    "try:\n",
    "    conn.execute(\"\"\"\n",
    "    DROP TABLE IF EXISTS sequence\n",
    "    \"\"\")\n",
    "    \n",
    "    conn.execute(\"\"\"\n",
    "    DROP TABLE IF EXISTS metadata\n",
    "    \"\"\")\n",
    "    \n",
    "    conn.execute(\"\"\"\n",
    "    DROP TABLE IF EXISTS phylogenetic\n",
    "    \"\"\")\n",
    "    \n",
    "    conn.execute(\"\"\"\n",
    "    DROP TABLE IF EXISTS source\n",
    "    \"\"\")\n",
    "    \n",
    "    conn.execute(\"\"\"\n",
    "    DROP TABLE IF EXISTS epitope\n",
    "    \"\"\")\n",
    "    \n",
    "    conn.execute(\"\"\"\n",
    "    CREATE TABLE IF NOT EXISTS sequence (assession TEXT, seq TEXT, PRIMARY KEY ('assession'));\n",
    "    \"\"\")\n",
    "    conn.execute(\"\"\"\n",
    "    CREATE TABLE IF NOT EXISTS metadata (assession TEXT, country TEXT, Year INT,\n",
    "                    FOREIGN KEY('assession') \n",
    "                    REFERENCES sequence ('assession')\n",
    "                    ON DELETE CASCADE\n",
    "                    ON UPDATE CASCADE);\n",
    "    \"\"\")\n",
    "    conn.execute(\"\"\"\n",
    "    CREATE TABLE IF NOT EXISTS ref (assession TEXT, title TEXT, journal TEXT,\n",
    "                    FOREIGN KEY('assession') \n",
    "                    REFERENCES sequence ('assession')\n",
    "                    ON DELETE CASCADE\n",
    "                    ON UPDATE CASCADE);  \n",
    "    \"\"\")\n",
    "    \n",
    "    conn.execute(\"\"\"\n",
    "    CREATE TABLE IF NOT EXISTS phylogenetic (assession TEXT, genotype TEXT,\n",
    "                    FOREIGN KEY('assession') \n",
    "                    REFERENCES sequence ('assession')\n",
    "                    ON DELETE CASCADE\n",
    "                    ON UPDATE CASCADE);  \n",
    "    \"\"\")\n",
    "    \n",
    "    conn.execute(\"\"\"\n",
    "    CREATE TABLE IF NOT EXISTS epitope (id TEXT, epitope TEXT,\n",
    "                    FOREIGN KEY('id') \n",
    "                    REFERENCES sequence ('id')\n",
    "                    ON DELETE CASCADE\n",
    "                    ON UPDATE CASCADE);  \n",
    "    \"\"\")\n",
    "    \n",
    "    conn.executemany(\"\"\"\n",
    "    INSERT INTO sequence VALUES (?, ?)\"\"\",[('AY911262','ATGGACTCTTACT'),\n",
    "                                           ('FJ614813','CTGGACGGTTTCC'),\n",
    "                                           ('FJ948820','GCTTTCGGGATAC'),\n",
    "                                           ('JF920046','ATGGACTCTTACT'),\n",
    "                                           ('JF920047','GCTTTCGGGATAC')])\n",
    "    conn.executemany(\"\"\"\n",
    "    INSERT INTO metadata VALUES (?,?,?)\"\"\",[('AY911262','USA',2007),\n",
    "                                           ('FJ614813','USA',2009),\n",
    "                                           ('FJ948820','Netherland',2010),\n",
    "                                           ('JF920046','USA',2012),\n",
    "                                           ('JF920047','USA',2009)])\n",
    "    conn.executemany(\"\"\"\n",
    "    INSERT INTO ref VALUES (?,?,?)\"\"\",[('AY911262','Respiratory syncytial virus nonstructural proteins NS1 and NS2','J. Virol'),\n",
    "                                      ('FJ614813','A chimeric A2 strain of respiratory syncytial virus','J. Virol'),\n",
    "                                      ('JF920047','Whole Genome Sequencing and Evolutionary','PLoS ONE 6')])\n",
    "    conn.executemany(\"\"\"\n",
    "    INSERT INTO phylogenetic VALUES (?, ?)\"\"\",[('FJ614813','ON1'),\n",
    "                                               ('FJ948820','GA2'),\n",
    "                                               ('JF920046','BA'),\n",
    "                                               ('JF920047','GB3')])\n",
    "    \n",
    "    conn.executemany(\"\"\"\n",
    "    INSERT INTO epitope VALUES (?, ?)\"\"\",[('AY911262','ATGGAC'),\n",
    "                                          ('FJ614813','CTGGAC'),\n",
    "                                          ('FJ948820','AGATAC'),\n",
    "                                          ('JF920046','ATGGACT'),\n",
    "                                          ('JF920047','GCTTTCG')])\n",
    "\n",
    "    cursor1 = conn.execute(\"select * from sequence\")\n",
    "    df1 = pd.DataFrame(cursor1.fetchall(),columns = [ 'assession','seq'] )              \n",
    "    print(df1)\n",
    "    \n",
    "    cursor2 = conn.execute('''select sequence.assession, sequence.seq, epitope.epitope from\n",
    "                           sequence join epitope\n",
    "                           on sequence.assession = epitope.id''')\n",
    "                           \n",
    "    df2 = pd.DataFrame(cursor2.fetchall(),columns = [ 'assession','seq','epitope'] )              \n",
    "    print(df2)\n",
    "    \n",
    "    cursor3 = conn.execute('''select sequence.assession, sequence.seq, metadata.country from\n",
    "                           sequence join metadata\n",
    "                           on sequence.assession = metadata.assession\n",
    "                           where country = 'USA' ''')\n",
    "    df3 = pd.DataFrame(cursor3.fetchall(),columns = [ 'assession','seq','country'] )              \n",
    "    print(df3)\n",
    "    \n",
    "    cursor4 = conn.execute('''select sequence.assession, sequence.seq, phylogenetic.genotype from \n",
    "                           sequence join phylogenetic\n",
    "                           on sequence.assession = phylogenetic.assession\n",
    "                           order by sequence.assession asc''')\n",
    "    df4 = pd.DataFrame(cursor4.fetchall(),columns = [ 'assession','seq','genotype'] )              \n",
    "    print(df4)\n",
    "    \n",
    "    cursor5 = conn.execute('''update phylogenetic set genotype = 'GA2' where genotype = 'ON1';''')\n",
    "    cursor5 = conn.execute('''select metadata.assession, metadata.country, phylogenetic.genotype from\n",
    "                           metadata join phylogenetic\n",
    "                           on metadata.assession = phylogenetic.assession''')\n",
    "    df5 = pd.DataFrame(cursor5.fetchall(),columns = [ 'assession','country','genotype'] )              \n",
    "    print(df5)                      \n",
    "                           \n",
    "                           \n",
    "    conn.commit();\n",
    "                           \n",
    "                           \n",
    "except sqlite3.Error as e:\n",
    "    print(\"There was error:\"+str(e))   \n",
    "    \n",
    "finally: \n",
    "    print(\"this excuted\")\n",
    "    conn.close()                  \n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
